File: scripts/provider_hetzner.js
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import axios from 'axios';
import dotenv from 'dotenv';
import { os } from 'node:os';

dotenv.config();

const HETZNER_API_URL = "https://api.hetzner.cloud/v1";

async function get_token() {
    return process.env.HCLOUD_TOKEN;
}

const client = (token) => axios.create({
    baseURL: HETZNER_API_URL,
    headers: {
        "Authorization": `Bearer ${token}`,
        "Content-Type": "application/json"
    }
});

async function setup_ssh_key(token) {
    const pubKeyPath = path.join(process.env.HOME || process.env.USERPROFILE, '.ssh', 'id_ed25519.pub');
    if (!fs.existsSync(pubKeyPath)) {
        console.error("SSH public key not found at " + pubKeyPath);
        return null;
    }

    const pubKey = fs.readFileSync(pubKeyPath, 'utf8').trim();
    const c = client(token);

    // Check if key already exists
    const keysRes = await c.get('ssh_keys');
    const existingKey = keysRes.data.ssh_keys.find(k => k.public_key.split(' ')[1] === pubKey.split(' ')[1]);

    if (existingKey) {
        return existingKey.id;
    }

    // Upload new key
    const res = await c.post('ssh_keys', {
        name: `agent-key-${Date.now()}`,
        public_key: pubKey
    });
    return res.data.ssh_key.id;
}

async function create_server(token, sshKeyId) {
    const c = client(token);
    try {
        const res = await c.post('servers', {
            name: `agent-node-${Date.now()}`,
            server_type: "cax11",
            image: "ubuntu-24.04",
            location: "nbg1",
            ssh_keys: [sshKeyId]
        });
        return res.data.server.id;
    } catch (error) {
        console.error("Error creating server:", error.response?.data || error.message);
        return null;
    }
}

async function wait_for_server(token, serverId) {
    const c = client(token);
    console.log(`Waiting for server ${serverId} to be ready...`);
    while (true) {
        const res = await c.get(`servers/${serverId}`);
        const status = res.data.server.status;
        if (status === "running") {
            return res.data.server.public_net.ipv4.ip;
        } else if (status === "off") {
            await c.post(`servers/${serverId}/actions/poweron`);
        }
        await new Promise(resolve => setTimeout(resolve, 5000));
    }
}

async function main() {
    const token = await get_token();
    if (!token) {
        console.error("Error: HCLOUD_TOKEN not found in .env or environment.");
        process.exit(1);
    }

    const keyId = await setup_ssh_key(token);
    if (!keyId) {
        console.error("Error: Failed to setup SSH key.");
        process.exit(1);
    }

    const serverId = await create_server(token, keyId);
    if (serverId) {
        const ip = await wait_for_server(token, serverId);
        console.log(`SUCCESS: Server is ready at ${ip}`);
        fs.writeFileSync("server_state.json", JSON.stringify({ server_id: serverId, ip: ip }, null, 2));
    }
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});

File: scripts/provider_hetzner.py
#!/usr/bin/env python3
import os
import sys
import requests
import json
import time

HETZNER_API_URL = "https://api.hetzner.cloud/v1"

def get_token():
    # Load from .env or environment
    if os.path.exists(".env"):
        with open(".env") as f:
            for line in f:
                if line.startswith("HCLOUD_TOKEN="):
                    return line.strip().split("=")[1]
    return os.getenv("HCLOUD_TOKEN")

def hetzner_post(endpoint, data, token):
    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }
    response = requests.post(f"{HETZNER_API_URL}/{endpoint}", json=data, headers=headers)
    return response.json()

def hetzner_get(endpoint, token):
    headers = {"Authorization": f"Bearer {token}"}
    response = requests.get(f"{HETZNER_API_URL}/{endpoint}", headers=headers)
    return response.json()

def setup_ssh_key(token):
    pub_key_path = os.path.expanduser("~/.ssh/id_ed25519.pub")
    if not os.path.exists(pub_key_path):
        print("SSH public key not found.")
        return None
    
    with open(pub_key_path) as f:
        pub_key = f.read().strip()
    
    # Check if key already exists
    keys = hetzner_get("ssh_keys", token)
    for k in keys.get("ssh_keys", []):
        if k["public_key"].split()[1] == pub_key.split()[1]:
            return k["id"]
    
    # Upload new key
    data = {
        "name": f"agent-key-{int(time.time())}",
        "public_key": pub_key
    }
    res = hetzner_post("ssh_keys", data, token)
    return res.get("ssh_key", {}).get("id")

def create_server(token, ssh_key_id):
    # Optimized for 2026: CAX11 (ARM64) is cost-effective
    data = {
        "name": f"agent-node-{int(time.time())}",
        "server_type": "cax11",
        "image": "ubuntu-24.04",
        "location": "nbg1", # Nuremberg
        "ssh_keys": [ssh_key_id]
    }
    res = hetzner_post("servers", data, token)
    if "error" in res:
        print(f"Error creating server: {res['error']}")
        return None
    return res["server"]["id"]

def wait_for_server(token, server_id):
    print(f"Waiting for server {server_id} to be ready...")
    while True:
        res = hetzner_get(f"servers/{server_id}", token)
        status = res.get("server", {}).get("status")
        if status == "running":
            ip = res["server"]["public_net"]["ipv4"]["ip"]
            return ip
        elif status == "off":
             # Try to power on if it's off
             hetzner_post(f"servers/{server_id}/actions/poweron", {}, token)
        
        time.sleep(5)

if __name__ == "__main__":
    token = get_token()
    if not token:
        print("Error: HCLOUD_TOKEN not found.")
        sys.exit(1)
    
    key_id = setup_ssh_key(token)
    if not key_id:
        print("Error: Failed to setup SSH key.")
        sys.exit(1)
    
    server_id = create_server(token, key_id)
    if server_id:
        ip = wait_for_server(token, server_id)
        print(f"SUCCESS: Server is ready at {ip}")
        # Save to a state file
        with open("server_state.json", "w") as f:
            json.dump({"server_id": server_id, "ip": ip}, f)
